packages/shared-ux/table_persist/src/use_table_persist.ts

```diff
@@ -0,0 +1,87 @@
+import { Criteria } from '@elastic/eui';
+import { DEFAULT_INITIAL_PAGE_SIZE, DEFAULT_PAGE_SIZE_OPTIONS } from './constants';
+import { createStorage } from './storage';
+import { validatePersistData } from './validate_persist_data';
+import { PersistData, PropertySort } from './types';
+
+export interface EuiTablePersistProps<T> {
+  /** A unique id that will be included in the local storage variable for this table. */
+  tableId: string;
+  /** (Optional) Specifies a custom onTableChange handler. */
+  customOnTableChange?: (change: Criteria<T>) => void;
+  /** (Optional) Specifies a custom initial table sorting. */
+  initialSort?: PropertySort<T>;
+  /** (Optional) Specifies a custom initial page size for the table. Defaults to 50. */
+  initialPageSize?: number;
+  /** (Optional) Specifies custom page size options for the table.
+   * Defaults to {@link DEFAULT_PAGE_SIZE_OPTIONS} */
+  pageSizeOptions?: number[];
+}
+
+/**
+ * A hook that stores and retrieves from local storage the table page size and sort criteria.
+ * Returns the persisting page size and sort and the onTableChange handler that should be passed
+ * as props to an Eui table component.
+ */
+export const useEuiTablePersist = <T extends object>({
+  tableId,
+  customOnTableChange,
+  initialSort,
+  initialPageSize,
+  pageSizeOptions,
+}: EuiTablePersistProps<T>) => {
+  const storage = createStorage();
+  const storedPersistData = storage.get(tableId, undefined);
+
+  const { pageSize: storagePageSize, sort: storageSort }: PersistData<T> = validatePersistData(
+    storedPersistData,
+    pageSizeOptions ?? DEFAULT_PAGE_SIZE_OPTIONS
+  );
+
+  const [pageSize, setPageSize] = useState<number>(
+    storagePageSize ?? initialPageSize ?? DEFAULT_INITIAL_PAGE_SIZE
+  );
+  const [sort, setSort] = useState<PropertySort<T> | undefined>(storageSort ?? initialSort);
+  const sorting = sort && { sort };
+
+  const onTableChange = useCallback(
+    (nextValues: Criteria<T>) => {
+      if (customOnTableChange) {
+        customOnTableChange(nextValues);
+      }
+
+      let nextSort: PropertySort<T> | undefined;
+      if (nextValues.sort?.field && nextValues.sort?.direction) {
+        // Both field and direction are needed for a valid sort criteria
+        nextSort = nextValues.sort;
+      } else if (!nextValues.sort?.field && !nextValues.sort?.direction) {
+        // If both field and direction are undefined, there is no change on sort so use the current one
+        nextSort = sort;
+      }
+
+      const nextPageSize = nextValues.page?.size ?? pageSize;
+
+      if (nextPageSize !== storagePageSize || nextSort !== storageSort) {
+        setPageSize(nextPageSize);
+        setSort(nextSort);
+        const nextPersistData: PersistData<T> = {
+          pageSize: nextPageSize,
+          sort: nextSort,
+        };
+        storage.set(tableId, nextPersistData);
+      }
+    },
+    [customOnTableChange, pageSize, sort, storage, storagePageSize, storageSort, tableId]
```

Comment:
In a callback that it setting the state (`setPageSize`, `setSort`), it is usually better to avoid having the state being set as a dependency. It could create an infinite loop. 
Here I checked and there are no infinite loop but I think we could avoid it this way

```ts
const onTableChange = useCallback(
  (nextValues: Criteria<T>) => {
    if (customOnTableChange) {
      customOnTableChange(nextValues);
    }

    let nextSort: PropertySort<T> | undefined;
    if (nextValues.sort?.field && nextValues.sort?.direction) {
      // Both field and direction are needed for a valid sort criteria
      nextSort = nextValues.sort;
    }

    if (nextValues.sort?.field || nextValues.sort?.direction) {
      setSort(nextSort);
    }

    const nextPageSize = nextValues.page?.size;
    if (nextPageSize) {
      setPageSize(nextPageSize);
    }

    if (
      (nextPageSize && nextPageSize !== storagePageSize) ||
      (nextSort && nextSort !== storageSort)
    ) {
      const nextPersistData: PersistData<T> = {
        pageSize: nextPageSize,
        sort: nextSort,
      };
      storage.set(tableId, nextPersistData);
    }
  },
  [customOnTableChange, storage, storagePageSize, storageSort, tableId]
);
```

Reply:
Good call! I added your suggestion with a small modification to ensure that sort would be cleaned up in local storage if the user removes the sort on a specific field (when this happens, `field` is an empty string). [7754207](https://github.com/elastic/kibana/pull/186997/commits/77542076080846eb27e852561a1e43f8f3b75af5)